components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/spm/spm_api.c

```diff
@@ -0,0 +1,328 @@
+#endif
+    if (g_spm_partition_db.partition_count >= SPM_MAX_PARTITIONS) {
+        return SPM_ERR_INVALID_CONFIG;
+    }
+    part_ptr = &(g_spm_partition_db.partitions[
+            g_spm_partition_db.partition_count]);
+    part_ptr->static_data.partition_id = TFM_SP_NON_SECURE_ID;
+    part_ptr->static_data.partition_flags = 0;
+
+#if TFM_LVL != 1
+    part_ptr->memory_data.stack_bottom = psp_stack_bottom;
+    part_ptr->memory_data.stack_top    = psp_stack_top;
+    /* Since RW, ZI and stack are configured as one MPU region, configure
+     * RW start address to psp_stack_bottom to get RW access to stack
+     */
+    part_ptr->memory_data.rw_start     = psp_stack_bottom;
+#endif
+
+    part_ptr->runtime_data.partition_state = SPM_PARTITION_STATE_UNINIT;
+    tfm_nspm_configure_clients();
+    ++g_spm_partition_db.partition_count;
+
+    /* For the TF-M core environment itself */
+    if (g_spm_partition_db.partition_count >= SPM_MAX_PARTITIONS) {
+        return SPM_ERR_INVALID_CONFIG;
+    }
+    part_ptr = &(g_spm_partition_db.partitions[
+            g_spm_partition_db.partition_count]);
+    part_ptr->static_data.partition_id = TFM_SP_CORE_ID;
+    part_ptr->static_data.partition_flags =
+                    SPM_PART_FLAG_APP_ROT | SPM_PART_FLAG_PSA_ROT;
+    part_ptr->runtime_data.partition_state = SPM_PARTITION_STATE_UNINIT;
+    ++g_spm_partition_db.partition_count;
+
+    /* Add user-defined secure partitions */
+    #include "tfm_partition_list.inc"
+
+    g_spm_partition_db.is_init = 1;
+
+    return SPM_ERR_OK;
+}
+
+enum spm_err_t tfm_spm_partition_init(void)
+{
+    struct spm_partition_desc_t *part;
+    struct tfm_sfn_req_s desc;
+    int32_t args[4] = {0};
+    int32_t fail_cnt = 0;
+    uint32_t idx;
+
+    /* Call the init function for each partition */
+    for (idx = 0; idx < g_spm_partition_db.partition_count; ++idx) {
+        part = &g_spm_partition_db.partitions[idx];
+#ifdef TFM_PSA_API
+        if (part->static_data.partition_flags & SPM_PART_FLAG_IPC) {
+            continue;
+        }
+#endif
+        tfm_spm_hal_configure_default_isolation(part->platform_data);
+        if (part->static_data.partition_init == NULL) {
+            tfm_spm_partition_set_state(idx, SPM_PARTITION_STATE_IDLE);
+            tfm_spm_partition_set_caller_partition_idx(idx,
+                                                     SPM_INVALID_PARTITION_IDX);
+        } else {
+            int32_t res;
+
+            desc.args = args;
+            desc.ns_caller = 0;
+            desc.sfn = (sfn_t)part->static_data.partition_init;
+            desc.sp_id = part->static_data.partition_id;
+            res = tfm_core_sfn_request(&desc);
+            if (res == TFM_SUCCESS) {
+                tfm_spm_partition_set_state(idx, SPM_PARTITION_STATE_IDLE);
+            } else {
+                tfm_spm_partition_err_handler(part, TFM_INIT_FAILURE, res);
+                fail_cnt++;
```

Comment:
Minor thing, but can't we just break here(right after fail_cnt++) instead of looping again? Or is it really required to call err_handler for each error?

Reply:
This file is imported from TFM repository (same as cmsis code), we added patches but they are mainly fixes for issues which were raised during the Musca port.